File: src/main/java/io/curiousoft/ijudi/ordermanagement/validator/ShippingValidator.java
package io.curiousoft.ijudi.ordermanagement.validator;

import io.curiousoft.ijudi.ordermanagement.model.BuildingType;
import io.curiousoft.ijudi.ordermanagement.model.Order;
import io.curiousoft.ijudi.ordermanagement.model.ShippingData;
import org.springframework.util.StringUtils;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;
import java.util.Date;

public class ShippingValidator implements ConstraintValidator<ValidDeliveryInfo, Order> {

        @Override
        public boolean isValid(Order order, ConstraintValidatorContext context) {
            ShippingData value = order.getShippingData();
            if(value == null) return true;
            boolean isValidBuildingType = value.getType() == ShippingData.ShippingType.COLLECTION
                    || value.getBuildingType() == BuildingType.HOUSE ||
                    (!StringUtils.isEmpty(value.getUnitNumber()) && !StringUtils.isEmpty(value.getBuildingName()));
            boolean isValidPickup = value.getType() == ShippingData.ShippingType.COLLECTION
                    && value.getPickUpTime() != null && value.getPickUpTime().after(new Date());
            boolean hasMessengerForDelivery = value.getMessengerId() != null;
            return isValidBuildingType && ( isValidPickup || hasMessengerForDelivery);
        }
}

File: izinga-commons/src/main/kotlin/commons/validator/ShippingValidator.kt
package io.curiousoft.izinga.commons.validator

import io.curiousoft.izinga.commons.model.BuildingType
import io.curiousoft.izinga.commons.model.Order
import io.curiousoft.izinga.commons.model.ShippingData
import org.springframework.util.StringUtils
import java.util.*
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

class ShippingValidator : ConstraintValidator<ValidDeliveryInfo?, Order> {
    override fun isValid(order: Order, context: ConstraintValidatorContext): Boolean {
        val value = order.shippingData ?: return true
        val isValidBuildingType =
            value.type == ShippingData.ShippingType.SCHEDULED_DELIVERY || value.buildingType == BuildingType.HOUSE ||
                    !StringUtils.isEmpty(value.unitNumber) && !StringUtils.isEmpty(value.buildingName)
        val isValidPickup =
            value.type == ShippingData.ShippingType.SCHEDULED_DELIVERY && value.pickUpTime?.after(
                Date()
            )?: false
        val hasMessengerForDelivery = value.messengerId != null
        return isValidBuildingType && (isValidPickup || hasMessengerForDelivery)
    }
}